### PR TITLE
fix: avoid spamming SyncPackage and InstallPackageRevision events

### LINKED ISSUE #4933: Functions get thousands of `InstallPackageRevision` events
### What happened?

We see these events a lot:
```bash
Events:
  Type    Reason                  Age                      From                                 Message
  ----    ------                  ----                     ----                                 -------
  Normal  InstallPackageRevision  4h56m (x2582 over 46h)   packages/function.pkg.crossplane.io  Successfully installed package revision
  Normal  InstallPackageRevision  2m22s (x297 over 4h52m)  packages/function.pkg.crossplane.io  Successfully installed package revision
```
The image used is [function-auto-ready](https://github.com/crossplane-contrib/function-auto-ready), but we built our self. How ever, this happens with all our functions.
Full describe:
```bash
❯ k describe function function-auto-ready
Name:         function-auto-ready
Namespace:
Annotations:  <none>
API Version:  pkg.crossplane.io/v1beta1
Kind:         Function
Metadata:
  Creation Timestamp:  2023-10-24T08:04:25Z
  Generation:          1
  Resource Version:    34054820
  UID:                 5f2210bc-56c3-4c61-853c-9416f01efad9
Spec:
  Ignore Crossplane Constraints:  false
  Package:                        registry.example.com/function-auto-ready-dev:latest
  Package Pull Policy:            Always
  Package Pull Secrets:
    Name:                      registry.example.com
  Revision Activation Policy:  Automatic
  Revision History Limit:      1
  Runtime Config Ref:
    API Version:               pkg.crossplane.io/v1beta1
    Kind:                      DeploymentRuntimeConfig
    Name:                      deployment-runtime-config
  Skip Dependency Resolution:  false
Status:
  Conditions:
    Last Transition Time:  2023-10-25T14:54:40Z
    Reason:                HealthyPackageRevision
    Status:                True
    Type:                  Healthy
    Last Transition Time:  2023-10-27T11:02:49Z
    Reason:                ActivePackageRevision
    Status:                True
    Type:                  Installed
  Current Identifier:      registry.example.com/function-auto-ready-dev:latest
  Current Revision:        function-auto-ready-3e23e6b1e41e
Events:
  Type    Reason                  Age                      From                                 Message
  ----    ------                  ----                     ----                                 -------
  Normal  InstallPackageRevision  5h1m (x2582 over 46h)    packages/function.pkg.crossplane.io  Successfully installed package revision
  Normal  InstallPackageRevision  2m59s (x302 over 4h58m)  packages/function.pkg.crossplane.io  Successfully installed package revision
```

```bash
❯ k get functionrevisions.pkg.crossplane.io
NAME                                HEALTHY   REVISION   IMAGE                                                    STATE      DEP-FOUND   DEP-INSTALLED   AGE
function-auto-ready-3e23e6b1e41e    True      1          registry.example.com/function-auto-ready-dev:latest   Active                                 6d11h
```
```bash
❯ k describe functionrevisions function-auto-ready-3e23e6b1e41e
Name:         function-auto-ready-3e23e6b1e41e
Namespace:
Labels:       pkg.crossplane.io/package=function-auto-ready
Annotations:  <none>
API Version:  pkg.crossplane.io/v1beta1
Kind:         FunctionRevision
Metadata:
  Creation Timestamp:  2023-10-24T08:04:27Z
  Finalizers:
    revision.pkg.crossplane.io
  Generation:  1
  Owner References:
    API Version:           pkg.crossplane.io/v1beta1
    Block Owner Deletion:  true
    Controller:            true
    Kind:                  Function
    Name:                  function-auto-ready
    UID:                   5f2210bc-56c3-4c61-853c-9416f01efad9
  Resource Version:        29986951
  UID:                     e02ba4a0-9e31-4dd7-a13e-4e3198341ebd
Spec:
  Desired State:                  Active
  Ignore Crossplane Constraints:  false
  Image:                          registry.example.com/function-auto-ready-dev:latest
  Package Pull Policy:            Always
  Package Pull Secrets:
    Name:    <redacted>
  Revision:  1
  Runtime Config Ref:
    API Version:               pkg.crossplane.io/v1beta1
    Kind:                      DeploymentRuntimeConfig
    Name:                      deployment-runtime-config
  Skip Dependency Resolution:  false
  Tls Server Secret Name:      <redacted>
Status:
  Conditions:
    Last Transition Time:  2023-10-25T14:54:39Z
    Reason:                HealthyPackageRevision
    Status:                True
    Type:                  Healthy
  Endpoint:                dns:///function-auto-ready.crossplane:9443
Events:
  Type    Reason       Age                       From                                         Message
  ----    ------       ----                      ----                                         -------
  Normal  SyncPackage  5h14m (x510749 over 46h)  packages/functionrevision.pkg.crossplane.io  Successfully configured package revision
  Normal  SyncPackage  6s (x63070 over 5h10m)    packages/functionrevision.pkg.crossplane.io  Successfully configured package revision
```

This happens with both:
```bash
apiVersion: pkg.crossplane.io/v1beta1
kind: Function
metadata:
  name: function-auto-ready
spec:
  runtimeConfigRef:
    name: deployment-runtime-config
  # package: xpkg.upbound.io/crossplane-contrib/function-auto-ready:v0.1.0
  package: registry.example.com/function-auto-ready-dev:latest
  packagePullPolicy: Always
  packagePullSecrets:
    - name: <redacted>
```
and

```bash
apiVersion: pkg.crossplane.io/v1beta1
kind: Function
metadata:
  name: function-auto-ready
spec:
  runtimeConfigRef:
    name: deployment-runtime-config
  package: function-auto-ready-dev@sha256:<image-sha-here>
  packagePullPolicy: Always
  packagePullSecrets:
    - name: <redacted>
```

### What environment did it happen in?
- `1.14.0-rc.0.611.g8121e6d6` and `1.14.0-rc.1` and `1.15.0-rc.0.16.g75b3b26e`
- Kubernetes Version: v1.26.7+c7ee51f
  - OpenShift

### PR DESCRIPTION
**Pull request by phisco**
### Description of your changes

Fixes https://github.com/crossplane/crossplane/issues/4933

I have: 

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [ ] ~Added or updated unit tests.~
- [ ] ~Added or updated e2e tests.~
- [ ] ~Linked a PR or a [docs tracking issue] to [document this change].~
- [x] Added `backport release-x.y` labels to auto-backport this PR.

Need help with this checklist? See the [cheat sheet].

[contribution process]: https://github.com/crossplane/crossplane/tree/master/contributing
[docs tracking issue]: https://github.com/crossplane/docs/issues/new
[document this change]: https://docs.crossplane.io/contribute/contribute
[cheat sheet]: https://github.com/crossplane/crossplane/tree/master/contributing#checklist-cheat-sheet

### CODE REVIEW

**Review by bobh66 — APPROVED** [review]
LGTM

File: internal/controller/pkg/manager/reconciler.go [hunk 1]
```diff
@@ -390,9 +390,13 @@ func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reco
 		}
 	}
 
-	if pr.GetCondition(v1.TypeHealthy).Status == corev1.ConditionTrue {
-		p.SetConditions(v1.Healthy())
+	// TODO(phisco): refactor these conditions to make it clearer
+	if pr.GetCondition(v1.TypeHealthy).Status == corev1.ConditionTrue && p.GetCondition(v1.TypeHealthy).Status != corev1.ConditionTrue {
+		// NOTE(phisco): We don't want to spam the user with events if the
+		// package is already healthy, nor update the condition last transition
+		// time if it didn't actually change.
```
Comment (negz) [inline internal/controller/pkg/manager/reconciler.go:397]: Hrm, I would not have thought we needed to do this. SetConditions should already be a no-op if we're setting a condition that already existed. 🤔

Comment (phisco) [reply internal/controller/pkg/manager/reconciler.go:397]: 🤔 you are right, I guess didn't get deep enough in the involved interfaces 😩 let me remove the comment and move back the condition setting out of the if

Comment (phisco) [reply internal/controller/pkg/manager/reconciler.go]: Pushed

**Review by jbw976 — COMMENTED** [review]
is this a regression from previous releases? 🤔

File: internal/controller/pkg/manager/reconciler.go [hunk 2]
```diff
@@ -390,9 +390,14 @@ func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reco
 		}
 	}
 
+	// TODO(phisco): refactor these conditions to make it clearer
 	if pr.GetCondition(v1.TypeHealthy).Status == corev1.ConditionTrue {
 		p.SetConditions(v1.Healthy())
```
Comment (negz) [inline internal/controller/pkg/manager/reconciler.go:395]: Should this be moved down below the following condition?

Right now it looks like we set the provider to healthy, then emit an event if the provider is not healthy. So I think the event would never be emitted. My guess is the intention is to omit the event only once, when the provider first becomes healthy. 😄

**Comment by negz** [comment]
> is this a regression from previous releases?

@jbw976 My _guess_ would be no. I suspect using pull: Always is uncommon, which is why we haven't noticed this before.

**Comment by phisco** [comment]
> is this a regression from previous releases? 🤔 

Nope, 1.13.2 emits 1 InstallPackageRevision per minute too with packagePullPolicy set to Always. 1.12 will too. I'll set backport labels for both of them.

File: internal/controller/pkg/manager/reconciler.go (hunk 2 above)
Comment (phisco) [reply internal/controller/pkg/manager/reconciler.go]: You are definitely right 🤦‍♂️

Comment (phisco) [reply internal/controller/pkg/manager/reconciler.go]: Done

**Comment by jbw976** [comment]
> I suspect using pull: Always is uncommon, which is why we haven't noticed this before.

My personal preference is to not take changes (and churn) this late before the release, esp. if the code is a common/hot path but it isn't affecting a common user scenarios. But the maintainer team can have final say about including in v1.14 or not 🙇

**Review by negz — APPROVED** [review]

**Comment by negz** [comment]
@phisco and I discussed with @jbw976 offline and I think we're in agreement that this is low-risk enough to merge (famous last words?). We do definitely agree with the sentiment WRT last minute changes though.

**Pull request merged**